README.md
# <img src="icons/icon-256.png" width="64px" align="center" alt="Unicopedia Plus icon"> UNICOPEDIA PLUS

**Unicopedia Plus** is a developer-oriented set of Unicode, Unihan & emoji utilities wrapped into one single app, built with [Electron](https://electronjs.org).

This app works on macOS, Linux and Windows operating systems.

## Utilities

The following utilities are currently available:

- **CJK Font Variants**
- **JavaScript Runner**
- **Regex Properties**
- **Emoji Data Finder**
    - **Find by Name**
    - **Match Sequence**
    - **Filter Text**
- **Emoji Picture Book**
- **Emoji References**
- **Unicode Data Finder**
    - **Find by Name**
    - **Match Character**
    - **List by Block**
- **Unicode Foldings**
- **Unicode Inspector**
- **Unicode Normalizer**
- **Unicode References**
- **Unihan Data Finder**
    - **Find by Tag Value**
    - **Match Character**
    - **View by Grid**
- **Unihan Inspector**
- **Unihan Radical-Strokes**
- **Unihan References**
- **Unihan Variants**

## CJK Font Variants

- The **CJK Font Variants** utility displays simultaneously any string of CJK (Chinese/Japanese/Korean) characters in five different typefaces belonging to the open-source set of [Google Noto CJK Fonts](https://github.com/googlefonts/noto-cjk):

| Language | Tag | Code | Typeface |
|----------|-----|------|----------|
| Japanese | ja | JP | Noto Sans CJK JP Regular |
| Korean | ko | KR | Noto Sans CJK KR Regular |
| Simplified Chinese | zh-Hans | SC | Noto Sans CJK SC Regular |
| Traditional Chinese (Taiwan) | zh-Hant-TW | TC | Noto Sans CJK TC Regular |
| Traditional Chinese (Hong Kong) | zh-Hant-HK | HK | Noto Sans CJK HK Regular |

- Additionally, it is possible to specify a set of logographic glyph variants for display by using the <kbd>East Asian Variant</kbd> drop-down menu.
- Font variants of the CJK characters can be visualized either vertically or horizontally. Use the <kbd>Writing Mode</kbd> drop-down menu to toggle between the two modes.
- Clicking inside any character frame displays momentarily the same glyph for all language flavors, while alt-clicking (or shift-clicking) applies to all characters of the string in a given language flavor. This is especially useful to quickly spot the differences between glyph variations. If the <kbd>⇪ Caps Lock</kbd> key is turned on as well, the differences are shown in contrasting colors instead.
- CJK characters can be entered either directly in the "Characters" input field, or using a series of code points in hexadecimal format in the "Code points" input field.
- It is also possible to input predefined strings of CJK characters selected from the <kbd>Samples&nbsp;▾</kbd> pop-up menu; some of them make use of the information found in the [StandardizedVariants.txt](https://www.unicode.org/Public/UNIDATA/StandardizedVariants.txt) or [IVD_Sequences.txt](https://www.unicode.org/ivd/data/2017-12-12/IVD_Sequences.txt) data files.
- As a convenience, the input fields can be emptied using the <kbd>Clear</kbd> button.
- In output, the standard Unicode code point format `U+7ADC` is used, i.e. "U+" directly followed by 4 or 5 hex digits.
- In input, more hexadecimal formats are allowed, including Unicode escape sequences, such as `\u9F8D` or `\u{20B9F}`. Moving out of the field or typing the Enter key converts all valid codes to standard Unicode code point format.

<img src="screenshots/cjk-font-variants-horizontal.png" width="1080px" alt="CJK Font Variants (Horizontal) screenshot">

<img src="screenshots/cjk-font-variants-vertical.png" width="1080px" alt="CJK Font Variants (Vertical) screenshot">

## JavaScript Runner

- The **JavaScript Runner** utility lets you execute JavaScript code, and comes with several sample scripts related to Unicode, Unihan and emoji; it is useful for quick testing/prototyping or data processing.

<img src="screenshots/javascript-runner.png" width="1080px" alt="JavaScript Runner screenshot">

## Regex Properties

- The **Regex Properties** utility displays all the **Unicode 12.1** properties available for regular expressions, used in particular in this app by the **Emoji Data Finder**, **Unicode Data Finder** and **Unihan Data Finder** utilities.
- These properties are suitable to build Unicode-aware regular expressions in JavaScript (ECMAScript 6) using the 'u' flag.
- Unicode properties fall into four groups, which can be displayed individually using the <kbd>Category</kbd> drop-down menu:
    - **General Category** properties
    - **Binary** properties
    - **Script** properties
    - **Script Extensions** properties
- For **General Category** properties, prefixing with `General_Category=` (Canonical) or `gc=` (Alias) is optional. Use the <kbd>Optional Prefix</kbd> checkbox to control whether the prefix is included or not.
- Groupings:

    | Property | Description |
    | -------- | ----------- |
    | Cased_Letter | Uppercase_Letter \| Lowercase_Letter \| Titlecase_Letter |
    | Letter | Uppercase_Letter \| Lowercase_Letter \| Titlecase_Letter \| Modifier_Letter \| Other_Letter |
    | Mark | Nonspacing_Mark \| Spacing_Mark \| Enclosing_Mark |
    | Number | Decimal_Number \| Letter_Number \| Other_Number |
    | Punctuation | Connector_Punctuation \| Dash_Punctuation \| Open_Punctuation \| Close_Punctuation \| Initial_Punctuation \| Final_Punctuation \| Other_Punctuation |
    | Symbol | Math_Symbol \| Currency_Symbol \| Modifier_Symbol \| Other_Symbol |
    | Separator | Space_Separator \| Line_Separator \| Paragraph_Separator |
    | Other | Control \| Format \| Surrogate \| Private_Use \| Unassigned |

- `\P{…}` is the negated form of `\p{…}`. Use the <kbd>Negated</kbd> checkbox to toggle between the two forms.
- Notes:
    - `\p{Any}` is equivalent to `[\u{0}-\u{10FFFF}]`
    - `\p{ASCII}` is equivalent to `[\u{0}-\u{7F}]`
    - `\p{Assigned}` is equivalent to `\P{Unassigned}` or `\P{General_Category=Unassigned}`

<img src="screenshots/regex-properties.png" width="1080px" alt="Regex Properties screenshot">

## Emoji Data Finder

### Find by Name

- The **Find by Name** feature of the **Emoji Data Finder** utility displays a list of basic data (symbol, short name, keywords, code points) of matching Unicode emoji searched by name or keyword, including through regular expressions.
- After entering a query, click on the <kbd>Search</kbd> button to display a list of all relevant matches, if any.
- This feature deals with the 4,022 emoji defined in the **Emoji 12.1** version of the [emoji-test.txt](https://www.unicode.org/Public/emoji/12.1/emoji-test.txt) data file.
- The 9 *component* (5 *skin-tone* and 4 *hair-style*) emoji and the 3,178 *fully-qualified* (**RGI**) emoji are presented in a standard way, while the 835 *non-fully-qualified* emoji are shown in a distinctive muted (grayed out) style.
- Use the <kbd>Results&nbsp;▾</kbd> pop-up menu to perform an action among:
    - `Copy Results` [copy the results as string to the clipboard]
    - `Save Results...` [save the results as string to a text file]
    - `Clear Results` [clear the current list of results]
- Various examples of regular expressions are provided for quick copy-and-paste.
- Note: **RGI** stands for *Recommended for General Interchange*.

<img src="screenshots/emoji-data-finder-find-by-name.png" width="1080px" alt="Emoji Data Finder - Find by Name screenshot">

### Match Sequence

- The **Match Sequence** feature of the **Emoji Data Finder** utility displays a list of basic data (symbol, short name, keywords, code points) of Unicode emoji matching a character sequence, including through regular expressions.
- After entering a query, click on the <kbd>Search</kbd> button to display a list of all relevant matches, if any.
- This feature deals with the 4,022 emoji defined in the **Emoji 12.1** version of the [emoji-test.txt](https://www.unicode.org/Public/emoji/12.1/emoji-test.txt) data file.
- The 9 *component* (5 *skin-tone* and 4 *hair-style*) emoji and the 3,178 *fully-qualified* (**RGI**) emoji are presented in a standard way, while the 835 *non-fully-qualified* emoji are shown in a distinctive muted (grayed out) style.
- Use the <kbd>Results&nbsp;▾</kbd> pop-up menu to perform an action among:
    - `Copy Results` [copy the results as string to the clipboard]
    - `Save Results...` [save the results as string to a text file]
    - `Clear Results` [clear the current list of results]
- Various examples of regular expressions are provided for quick copy-and-paste.
- Note: **RGI** stands for *Recommended for General Interchange*.

<img src="screenshots/emoji-data-finder-match-sequence.png" width="1080px" alt="Emoji Data Finder - Match Sequence screenshot">

### Filter Text

- The **Filter Text** feature of the **Emoji Data Finder** utility displays in real time a list of basic data (symbol, short name, keywords, code points) of all the Unicode emoji contained in a text string.
- Text can by directly typed or pasted from the clipboard into the main input field.
- It is also possible to input predefined sets of emoji selected from the <kbd>Samples&nbsp;▾</kbd> pop-up menu.
- As a convenience, the input field can be emptied using the <kbd>Clear</kbd> button.
- Use the <kbd>Filter&nbsp;▾</kbd> pop-up menu to perform an action among:
    - `Discard Non-Emoji` [strip out non-emoji characters]
    - `Upgrade to RGI Emoji` [restore incomplete emoji to their **RGI** form]
    - `Remove Duplicate Emoji` [delete emoji duplicates]
- This feature deals with the 4,022 emoji defined in the **Emoji 12.1** version of the [emoji-test.txt](https://www.unicode.org/Public/emoji/12.1/emoji-test.txt) data file.
- The 9 *component* (5 *skin-tone* and 4 *hair-style*) emoji and the 3,178 *fully-qualified* (**RGI**) emoji are presented in a standard way, while the 835 *non-fully-qualified* emoji are shown in a distinctive muted (grayed out) style.
- Use the <kbd>Results&nbsp;▾</kbd> pop-up menu to perform an action among:
    - `Copy Results` [copy the results as string to the clipboard]
    - `Save Results...` [save the results as string to a text file]
- Note: **RGI** stands for *Recommended for General Interchange*.

<img src="screenshots/emoji-data-finder-filter-text.png" width="1080px" alt="Emoji Data Finder - Filter Text screenshot">

## Emoji Picture Book

- The **Emoji Picture Book** utility displays lists of Unicode emoji in a color picture book fashion.
- Any group of pictures can be displayed by selecting its name in the <kbd>Category</kbd> drop-down menu, among:
    - **Smileys & Emotion**
    - **People & Body**
    - **Component**
    - **Animals & Nature**
    - **Food & Drink**
    - **Travel & Places**
    - **Activities**
    - **Objects**
    - **Symbols**
    - **Flags**
- The size of all emoji pictures (from 32 to 128&nbsp;pixels) can be adjusted by moving the dedicated slider left and right.
- The groups and subgroups of emoji are those defined in the **Emoji 12.1** version of the [emoji-test.txt](https://www.unicode.org/Public/emoji/12.1/emoji-test.txt) data file.
- Only the 9 *component* emoji and the 3,178 *fully-qualified* (**RGI**) forms of the emoji are used unless they cannot be displayed properly, depending on the emoji support level of the operating system.
- Emoji failing to be represented as proper color pictures are purely and simply discarded.
- Note: **RGI** stands for *Recommended for General Interchange*.

<img src="screenshots/emoji-picture-book.png" width="1080px" alt="Emoji Picture Book screenshot">

## Emoji References


CHANGES.md
# Release Notes

This project adheres to [Semantic Versioning](https://semver.org/).

## 6.2.2

- Fixed missing handling of `Save Results...` in the `Unihan Radical-Strokes` utility.

## 6.2.1

- Allowed alt-click as an alternative to shift-click in the `CJK Font Variants` utility.
- Added a hidden "save DOT source" debug feature to the `Unihan Variants` utility.
- Updated instructions.

## 6.2.0

- Added the full list of radical-strokes combinations as tooltip to each Unihan character listed in the `Unihan Radical-Strokes` utility.
- Added the detailed list of code points with Unicode names as tooltip to each emoji listed in the `Emoji Data Finder` utility.

## 6.1.0

- Added a "Save as SVG" feature to the `Unihan Variants` utility.
- Updated instructions and screenshots accordingly.

## 6.0.0

- Added a new `Unihan Variants` utility.
- Updated screenshots, instructions and `README.md` file accordingly.
- Improved horizontal centering of Unihan characters in the `View by Grid` feature of the `Unihan Data Finder` utility.
- Enabled opening of external links in SVG graphs.
- Updated `Electron` to version `7.1.9`.
- Updated `Electron Packager` to version `14.2.0`.

## 5.16.5

- Fixed opening of invalid anchors.
- Updated `Electron` to version `7.1.8`.

## 5.16.4

- Fixed app timestamp date.
- Updated copyright years.
- Updated emoji version dates.
- Updated handling of default smart zoom setting.

## 5.16.3

- Improved display of data table in the **Match Character** feature of the **Unihan Data Finder** utility, by graying out rows of variant characters, if any.
- Increased font size of Unihan input field for better legibility.
- Updated screenshot.

## 5.16.2

- Improved layout of Unihan data tables, by refactoring styling code.
- Cleaned up variable and class names in Unihan data tables code.
- Added regex example.
- Updated screenshot.
- Updated `Electron` to version `7.1.7`.

## 5.16.1

- Used constant terminology for characters and code points.
- Fixed typo.
- Simplified code.
- Added regex example.
- Updated screenshots.

## 5.16.0

- Moved the **Radical/Strokes** feature out of the **Unihan Data Finder** utility to a new **Unihan Radical-Strokes** utility.
- Added a new **Match Character** feature to the **Unihan Data Finder** utility, including a `Match Variants` option.
- Updated instructions and screenshots accordingly.
- Updated `Electron` to version `7.1.6`.

## 5.15.3

- Improved compact layout of chart of the **Radical/Strokes** feature of the **Unihan Data Finder** utility.
- Improved unique names for surrogates and private use characters, making use of a pattern consistent with the Unicode name derivation rules.
- Fixed removal of menu bar in license window on Linux.
- Updated `Electron` to version `7.1.5`.

## 5.15.2

- Fixed unique names of CJK Unified Ideographs, making use of the Unicode name derivation rules.
- Updated screenshots accordingly.
- Added samples for unassigned characters to the **Unicode Inspector** utility.
- Updated `Electron` to version `7.1.4`.

## 5.15.1

- Fixed incorrect display of age information in specific cases.
- Fixed missing non-character information for plane 16 private use characters.
- Added regex example to the **Match Character** feature of the **Unicode Data Finder** utility, for the total of 137,929 Unicode characters defined in [Unicode 12.1](https://unicode.org/versions/Unicode12.1.0/).

## 5.15.0

- Added tooltips with age, script, script extensions and general category to the data table rows of the **Unicode Data Finder** utility.
- Added tooltips with age and status of Unihan characters (unified or compatibility ideograph) to the data table rows of the **Find by Tag Value** feature, and to the data table cells of the **View by Grid** feature of the **Unihan Data Finder** utility.
- Added status (unified or compatibility ideograph) to the basic Unihan information of the **Unihan Inspector** utility.
- Appended date (month & year) to the age information in all utilities.
- Fixed missing age information for plane 16 private use characters.
- Updated instructions and screenshots accordingly.

## 5.14.0

- Replaced deprecated HTML Imports with explicit reading of HTML files and dynamic templates.
- Updated `Electron` to version `7.1.3`.

## 5.13.2

- Renamed license HTML file to prevent licensing ambiguity on GitHub.
- Added appropriate `lang` attribute to CJK reference links.
- Replaced custom code with `padStart()` string function for cleaner code and better performance.
- Fixed UI init order issue and added tooltip to each option of the `Locale` drop-down menu in the **Unicode Foldings** utility.

## 5.13.1

- Used stronger contrasting colors for the diff color mode of the **CJK Font Variants** utility.
- Added new regex example for `kTotalStrokes` to the **Find by Tag Value** feature of the **Unihan Data Finder** utility.
- Updated `Electron Packager` to version `14.1.1`.

## 5.13.0

- Added diff color mode to the **CJK Font Variants** utility.
- Cleaned up table creation in the **Radical/Strokes** feature of the **Unihan Data Finder** utility.
- Updated various samples.
- Updated `Electron` to version `7.1.2`.

## 5.12.1

- Added new CJK samples.
- Improved support for `Emoji 12.1`.
- Added separator lines for non compact layout in the `Radical/Strokes` feature of the `Unihan Data Finder` utility.
- Updated `Electron` to version `7.1.1`.

## 5.12.0

- Improved layout of checkboxes.
- Added support for `Emoji 12.1`.
- Added `Display Emoji Size Statistics` sample script.
- Updated `Electron` to version `7.1.0`.

## 5.11.1

- Fixed issues in `<input type="search">` fields by reverting `Electron` to version `6.1.1`.

## 5.11.0


renderer/units/Unihan Variants/import.js
//
const unit = document.getElementById ('unihan-variants-unit');
//
const unihanInput = unit.querySelector ('.unihan-input');
const lookupButton = unit.querySelector ('.lookup-button');
const extraVariantsCheckbox = unit.querySelector ('.extra-variants-checkbox');
const detailedRelationsCheckbox = unit.querySelector ('.detailed-relations-checkbox');
const linearCharacter = unit.querySelector ('.linear-character');
const linearVariants = unit.querySelector ('.linear-variants');
const saveSVGButton = unit.querySelector ('.save-svg-button');
const graphContainer = unit.querySelector ('.graph-container');
//
const instructions = unit.querySelector ('.instructions');
//
const unihanHistorySize = 256;   // 0: unlimited
//
let unihanHistory = [ ];
let unihanHistoryIndex = -1;
let unihanHistorySave = null;
//
let currentUnihanCharacter;
//
let defaultFolderPath;
//
module.exports.start = function (context)
{
    const { remote } = require ('electron');
    //
    const fs = require ('fs');
    const path = require ('path');
    //
    const fileDialogs = require ('../../lib/file-dialogs.js');
    //
    // https://github.com/mdaines/viz.js/wiki/Usage
    // https://github.com/mdaines/viz.js/wiki/Caveats
    //
    const Viz = require ('viz.js');
    const { Module, render } = require ('viz.js/full.render.js');
    //
    let viz = new Viz ({ Module, render });
    //
    const dotTemplate = fs.readFileSync (path.join (__dirname, 'template.dot'), { encoding: 'utf8' });
    //
    const unicode = require ('../../lib/unicode/unicode.js');
    const unihanData = require ('../../lib/unicode/parsed-unihan-data.js');
    const yasuokaVariants = require ('../../lib/unicode/parsed-yasuoka-variants-data.js');
    //
    const defaultPrefs =
    {
        unihanHistory: [ ],
        unihanCharacter: "",
        extraVariantsCheckbox: false,
        detailedRelationsCheckbox: false,
        defaultFolderPath: context.defaultFolderPath,
        instructions: true
    };
    let prefs = context.getPrefs (defaultPrefs);
    //
    unihanHistory = prefs.unihanHistory;
    //
    // Unihan characters
    const unihanPattern = '(?:(?=\\p{Script=Han})(?=\\p{Other_Letter}).)';
    const unihanRegex = new RegExp (unihanPattern, 'u');
    //
    const unifiedPattern = '\\p{Unified_Ideograph}';
    const unifiedRegex = new RegExp (unifiedPattern, 'u');
    //
    const characterOrCodePointRegex = /^\s*(?:(.)|(?:[Uu]\+)?([0-9a-fA-F]{4,5}|10[0-9a-fA-F]{4}))\s*$/u;
    //
    function parseUnihanCharacter (inputString)
    {
        let character = "";
        let match = inputString.match (characterOrCodePointRegex);
        if (match)
        {
            if (match[1])
            {
                character = match[1];
            }
            else if (match[2])
            {
                character = String.fromCodePoint (parseInt (match[2], 16));
            }
            if (!unihanRegex.test (character))
            {
                character = "";
            }
        }
        return character;
    }
    // 
    unihanInput.addEventListener
    (
        'input',
        (event) =>
        {
            event.currentTarget.classList.remove ('invalid');
            if (event.currentTarget.value)
            {
                if (!parseUnihanCharacter (event.currentTarget.value))
                {
                    event.currentTarget.classList.add ('invalid');
                }
            }
        }
    );
    unihanInput.addEventListener
    (
        'keypress',
        (event) =>
        {
            if (event.key === 'Enter')
            {
                event.preventDefault ();
                lookupButton.click ();
            }
        }
    );
    unihanInput.addEventListener
    (
        'keydown',
        (event) =>
        {
            if (event.altKey)
            {
                if (event.key === 'ArrowUp')
                {
                    event.preventDefault ();
                    if (unihanHistoryIndex === -1)
                    {
                        unihanHistorySave = event.currentTarget.value;
                    }
                    unihanHistoryIndex++;
                    if (unihanHistoryIndex > (unihanHistory.length - 1))
                    {
                        unihanHistoryIndex = (unihanHistory.length - 1);
                    }
                    if (unihanHistoryIndex !== -1)
                    {
                        event.currentTarget.value = unihanHistory[unihanHistoryIndex];
                        event.currentTarget.dispatchEvent (new Event ('input'));
                    }
                }
                else if (event.key === 'ArrowDown')
                {
                    event.preventDefault ();
                    unihanHistoryIndex--;
                    if (unihanHistoryIndex < -1)
                    {
                        unihanHistoryIndex = -1;
                        unihanHistorySave = null;
                    }
                    if (unihanHistoryIndex === -1)
                    {
                        if (unihanHistorySave !== null)
                        {
                            event.currentTarget.value = unihanHistorySave;
                            event.currentTarget.dispatchEvent (new Event ('input'));
                        }
                    }
                    else
                    {
                        event.currentTarget.value = unihanHistory[unihanHistoryIndex];
                        event.currentTarget.dispatchEvent (new Event ('input'));
                    }
                }
            }
        }
    );
    //
    function getTooltip (character)
    {
        let data = unicode.getCharacterBasicData (character);
        let status = unifiedRegex.test (character) ? "Unified Ideograph" : "Compatibility Ideograph";
        return `Code Point: ${data.codePoint}\nAge: Unicode ${data.age} (${data.ageDate})\nStatus: ${status}`;
    }
    //
    let variantTags =
    [
        'kCompatibilityVariant',
        'kSemanticVariant',
        'kSimplifiedVariant',
        'kSpecializedSemanticVariant',
        'kTraditionalVariant',
        'kZVariant'
    ];
    //
    function getVariantRelations (character)
    {
        let relations = [ ];
        for (let codePoint of unihanData.fullSet)
        {
            let setCharacter = String.fromCodePoint (parseInt (codePoint.replace ("U+", ""), 16));
            let codePointData = unihanData.codePoints[codePoint];
            for (let variantTag of variantTags)
            {
                if (variantTag in codePointData)
                {
                    let variants = codePointData[variantTag];
                    if (!Array.isArray (variants))
                    {
                        variants = [ variants ];
                    }
                    for (let variant of variants)
                    {
                        variant = variant.split ("<")[0];
                        let variantCharacter = String.fromCodePoint (parseInt (variant.replace ("U+", ""), 16));
                        if (setCharacter === character)
                        {
                            relations.push ({ to: variantCharacter, tag: variantTag });
                        }
                        else if (character === variantCharacter)
                        {
                            relations.push ({ from: setCharacter, tag: variantTag });
                        }
                    }
                }
            }
            if (extraVariantsCheckbox.checked)
            {
                // Yasuoka Variants
                let variants = yasuokaVariants[codePoint] || [ ];
                for (let variant of variants)
                {
                    if (variant in unihanData.codePoints)
                    {
                        let variantCharacter = String.fromCodePoint (parseInt (variant.replace ("U+", ""), 16));
                        if (setCharacter === character)
                        {
                            relations.push ({ to: variantCharacter, tag: 'kYasuokaVariant' });
                        }
                        else if (character === variantCharacter)
                        {
                            relations.push ({ from: setCharacter, tag: 'kYasuokaVariant' });
                        }
                    }
                }
            }
        }
        return relations;
    }
    //
    let shortLabels =
    {
        'kCompatibilityVariant': " Unified ",
        'kSemanticVariant': " Semantic ",
        'kSimplifiedVariant': " Simplified ",
        'kSpecializedSemanticVariant': " Specialized ",
        'kTraditionalVariant': " Traditional ",
        'kZVariant': " Shape ",
        'kYasuokaVariant': " Yasuoka "
    };
    //
    let dotString;
    let svgResult;
    //
    function displayData (character)
    {
        dotString = "";
        svgResult = "";
        saveSVGButton.disabled = true;
        // linearCharacter.textContent = "";
        while (linearCharacter.firstChild)
        {
            linearCharacter.firstChild.remove ();
        }
        // linearVariants.textContent = "";
        while (linearVariants.firstChild)
        {
            linearVariants.firstChild.remove ();
        }
        while (graphContainer.firstChild)
        {
            graphContainer.firstChild.remove ();
        }
        currentUnihanCharacter = character;
        if (character)
        {
            let indexOfUnihanCharacter = unihanHistory.indexOf (character);
            if (indexOfUnihanCharacter !== -1)
            {
                unihanHistory.splice (indexOfUnihanCharacter, 1);
            }
            unihanHistory.unshift (character);
            if ((unihanHistorySize > 0) && (unihanHistory.length > unihanHistorySize))
            {
                unihanHistory.pop ();
            }
            unihanHistoryIndex = -1;
            unihanHistorySave = null;
            let symbols = document.createElement ('div');
            symbols.className = 'symbols';
            let symbol = document.createElement ('span');
            symbol.className = 'symbol';
            symbol.title = getTooltip (character);
            symbol.textContent = character;
            symbols.appendChild (symbol);
            linearCharacter.appendChild (symbols);
            let relations = getVariantRelations (character);
            // console.log (relations);
            let variants = [ ];
            for (let relation of relations)
            {
                if ("to" in relation)
                {
                    variants.push (relation["to"]);
                }
                else if ("from" in relation)
                {
                    variants.push (relation["from"]);
                }
            }
            variants = [...new Set (variants.filter (variant => variant !== character))].sort ((a, b) => a.codePointAt (0) - b.codePointAt (0));
            // console.log (variants);
            if (variants.length > 0)
            {
                let symbols = document.createElement ('div');
                symbols.className = 'symbols';
                for (let variant of variants)
                {
                    let symbol = document.createElement ('span');
                    symbol.className = 'symbol';
                    symbol.title = getTooltip (variant);
                    symbol.textContent = variant;
                    symbols.appendChild (symbol);
                }
                linearVariants.appendChild (symbols);
            }
            try
            {
                let data = "";
                data += `    "${character}" [ tooltip = ${JSON.stringify (getTooltip (character))}, style = bold ]`;
                data += variants.map (variant => `\n    "${variant}" [ tooltip = ${JSON.stringify (getTooltip (variant))} ]`).join ("");
                if (detailedRelationsCheckbox.checked)
                {
                    let toCharacters = { };
                    let fromCharacters = { };
                    for (let relation of relations)
                    {
                        if ("to" in relation)
                        {
                            let toCharacter = relation["to"];
                            if (!(toCharacter in toCharacters))
                            {
                                toCharacters[toCharacter] = [ ];
                            }
                            toCharacters[toCharacter].push (relation["tag"]);
                        }
                        else if ("from" in relation)
                        {
                            let fromCharacter = relation["from"];
                            if (!(fromCharacter in fromCharacters))
                            {
                                fromCharacters[fromCharacter] = [ ];
                            }
                            fromCharacters[fromCharacter].push (relation["tag"]);
                        }
                    }
                    // console.log (toCharacters, fromCharacters);
                    for (let toCharacter in toCharacters)
                    {
                        let tags = toCharacters[toCharacter];
                        for (let tag of tags)
                        {
                            if ((tag !== 'kYasuokaVariant') || (tags.length === 1))
                            {
                                data += `\n    "${character}" -> "${toCharacter}" [ label = ${JSON.stringify (shortLabels[tag])} ]`;
                            }
                        }
                    }
                    for (let fromCharacter in fromCharacters)
                    {
                        let tags = fromCharacters[fromCharacter];
                        for (let tag of tags)
                        {
                            if ((tag !== 'kYasuokaVariant') || (tags.length === 1))
                            {
                                data += `\n    "${fromCharacter}" -> "${character}" [ label = ${JSON.stringify (shortLabels[tag])} ]`;
                            }
                        }
                    }
                }
                else
                {
                    data += variants.map (variant => `\n    "${character}" -- "${variant}"`).join ("");
                }
                // console.log (data);
                dotString =
                    dotTemplate
                    .replace ('{{graph}}', detailedRelationsCheckbox.checked ? 'digraph' : 'graph')
                    .replace ('{{rankdir}}', detailedRelationsCheckbox.checked ? 'LR' : 'TB')
                    .replace ('{{data}}', data);
                // console.log (dotString);
                viz.renderString (dotString, { engine: 'dot', format: 'svg' })
                .then
                (
                    result =>
                    {
                        svgResult = result;
                        graphContainer.innerHTML = svgResult;
                        saveSVGButton.disabled = false;
                    }
                );
            }
            catch (e)
            {
            }
        }
    }
    //
    function updateUnihanData (character)
    {
        unihanInput.value = "";
        unihanInput.dispatchEvent (new Event ('input'));
        displayData (character);
        unit.scrollTop = 0;
    }
    //
    graphContainer.addEventListener
    (
        'click',
        (event) =>
        {
            let aTag = event.target.closest ('a');
            if (aTag)
            {
                event.preventDefault ();
                let character = aTag.querySelector ('text').textContent;
                if (character != currentUnihanCharacter)
                {
                    updateUnihanData (character);
                }
            }
        }
    );
    //
    lookupButton.addEventListener
    (
        'click',
        (event) =>
        {
            if (unihanInput.value)
            {
                let character = parseUnihanCharacter (unihanInput.value);
                if (character)
                {
                    updateUnihanData (character);
                }
                else
                {
                    remote.shell.beep ();
                }
            }
            else
            {
                unihanHistoryIndex = -1;
                unihanHistorySave = null;
                updateUnihanData ("");
            }
        }
    );
    //
    extraVariantsCheckbox.checked = prefs.extraVariantsCheckbox;
    extraVariantsCheckbox.addEventListener ('click', (event) => { updateUnihanData (currentUnihanCharacter); });
    //
    detailedRelationsCheckbox.checked = prefs.detailedRelationsCheckbox;
    detailedRelationsCheckbox.addEventListener ('click', (event) => { updateUnihanData (currentUnihanCharacter); });
    //
    currentUnihanCharacter = prefs.unihanCharacter;
    updateUnihanData (currentUnihanCharacter);
    //
    defaultFolderPath = prefs.defaultFolderPath;
    //
    saveSVGButton.addEventListener
    (
        'click',
        (event) =>
        {
            let dotFormat = event.altKey || event.shiftKey;
            if (dotFormat)
            {
                fileDialogs.saveTextFile
                (
                    "Save DOT file:",
                    [ { name: "DOT File (*.dot)", extensions: [ 'dot' ] } ],
                    path.join (defaultFolderPath, `${currentUnihanCharacter}-Variants.dot`),
                    (filePath) =>
                    {
                        defaultFolderPath = path.dirname (filePath);
                        return dotString;
                    }
                );
            }
            else
            {
                fileDialogs.saveTextFile
                (
                    "Save SVG file:",
                    [ { name: "SVG File (*.svg)", extensions: [ 'svg' ] } ],
                    path.join (defaultFolderPath, `${currentUnihanCharacter}-Variants.svg`),
                    (filePath) =>
                    {
                        defaultFolderPath = path.dirname (filePath);
                        return svgResult;
                    }
                );
            }
        }
    );
    //
    instructions.open = prefs.instructions;
};
//
module.exports.stop = function (context)
{
    let prefs =
    {
        unihanHistory: unihanHistory,
        unihanCharacter: currentUnihanCharacter,
        extraVariantsCheckbox: extraVariantsCheckbox.checked,
        detailedRelationsCheckbox: detailedRelationsCheckbox.checked,
        defaultFolderPath: defaultFolderPath,
        instructions: instructions.open
    };
    context.setPrefs (prefs);
};
//
